flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSourceReaderMetricGroup.java

```diff
@@ -0,0 +1,143 @@
+@Internal
+public class InternalSourceReaderMetricGroup extends ProxyMetricGroup<MetricGroup>
+        implements SourceReaderMetricGroup {
+
+    public static final long ACTIVE = Long.MAX_VALUE;
+    private final OperatorIOMetricGroup operatorIOMetricGroup;
+    private final Clock clock;
+    private final Counter numRecordsInErrors;
+    // only if source emits at least one watermark
+    private SettableGauge<Long> watermarkGauge;
+    // only if records with event timestamp are emitted
+    private SettableGauge<Long> eventTimeGauge;
+    private long idleStartTime = ACTIVE;
+
+    private InternalSourceReaderMetricGroup(
+            MetricGroup parentMetricGroup,
+            OperatorIOMetricGroup operatorIOMetricGroup,
+            Clock clock) {
+        super(parentMetricGroup);
+        numRecordsInErrors = parentMetricGroup.counter(MetricNames.NUM_RECORDS_IN_ERRORS);
+        this.operatorIOMetricGroup = operatorIOMetricGroup;
+        this.clock = clock;
+        parentMetricGroup.gauge(
+                MetricNames.SOURCE_IDLE_TIME_GAUGE,
+                () -> isIdling() ? this.clock.absoluteTimeMillis() - idleStartTime : 0);
+    }
+
+    public static InternalSourceReaderMetricGroup wrap(OperatorMetricGroup operatorMetricGroup) {
+        return new InternalSourceReaderMetricGroup(
+                operatorMetricGroup,
+                operatorMetricGroup.getIOMetricGroup(),
+                SystemClock.getInstance());
+    }
+
+    @VisibleForTesting
+    public static InternalSourceReaderMetricGroup mock(MetricGroup metricGroup) {
+        return new InternalSourceReaderMetricGroup(
+                metricGroup,
+                UnregisteredMetricsGroup.createUnregisteredOperatorIOMetricGroup(),
+                SystemClock.getInstance());
+    }
+
+    private boolean isIdling() {
+        return idleStartTime != ACTIVE;
+    }
+
+    public void idlingStarted() {
+        if (!isIdling()) {
+            idleStartTime = clock.absoluteTimeMillis();
+        }
+    }
+
+    public void recordEmitted() {
+        idleStartTime = ACTIVE;
+    }
+
+    @Override
+    public Counter getNumRecordsInErrorsCounter() {
+        return numRecordsInErrors;
+    }
+
+    public void watermarkEmitted(long watermark) {
+        // iff a respective source emits a watermark, Flink can provide the watermark lag
+        if (watermarkGauge == null) {
+            watermarkGauge = new SettableGauge<>(watermark);
+            parentMetricGroup.gauge(
+                    MetricNames.WATERMARK_LAG,
+                    () -> clock.absoluteTimeMillis() - watermarkGauge.getValue());
+        } else {
+            watermarkGauge.setValue(watermark);
+        }
+    }
+
+    public void eventTimeEmitted(long timestamp) {
+        // iff a respective source emits a timestamp, Flink can provide the event lag
+        if (eventTimeGauge == null) {
+            eventTimeGauge = new SettableGauge<>(timestamp);
```

Comment:
What exactly does the SettableGauge provide us here opposed to a plain `long` that defaults to -1?

Reply:
Gotcha.